docs/how-to/internal/x11passthrough.rst

```diff
@@ -0,0 +1,270 @@
+``--add-host``, ``host.docker.internal`` is mapped to the keyword
+``host-gateway``. This keyword is a special value that the Docker server will
+replace with an address from which the host will be reachable within a
+container whether it is Docker Engine or Docker Desktop starting it.
+
+In conclusion, we can add ``--add-host host.docker.internal:host-gateway`` to
+the options to start a container and the host network interface will be
+reachable at ``host.docker.internal``.
+
+Exposing an X Display to a Container
+------------------------------------
+
+Given the knowledge of the operation of the Docker implementations, we finally
+have the pieces to expose an X display to a container. Since it is not possible
+to expose the display's socket directly to a container, a TCP proxy is
+configured to pass X messages on the network from the container to the socket
+on the host machine for the display.
+
+TCP Proxy
+~~~~~~~~~
+
+The `socat <https://repo.or.cz/socat.git>`__ tool is a widely available program
+to relay bi-directional data transfer between independent data channels. It
+allows running a process on the host to listen on a network port and send any
+received data to a socket connected to an X display on the other side.
+
+Creating a TCP proxy for the X display effectively creates a spoofed X display.
+The proxy is configured to listen on the TCP port for an unallocated display;
+the port number will be 6000 + the number of the display. Additionally, the
+proxy is configured to listen on all network interfaces since identifying the
+exact interface that will be available within the container is non-trivial.
+
+The other side of the proxy is connected to the socket for the X display. The
+socket, though, for the display may actually be another TCP socket; this will
+be the case if the environment is currently configured for X11 forwarding over
+SSH, as discussed below. In most cases, though, the socket will be the UNIX
+socket for the display in the ``/tmp/.X11-unix/`` directory.
+
+X Authentication for the Proxied X Display
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Authentication for X displays is managed in ``xauth`` database files. The
+``xauth`` program allows for reading and writing the database. The database is
+relatively simple mapping of display information to a cookie. When a client
+wants to establish a connection for a display, it queries the database for the
+display and receives the cookie back.
+
+Since the proxy creates a spoofed display, a new ``xauth`` database needs to be
+created for the spoofed display using the authentication afforded to the user
+for the current display.
+
+Steps to create new database:
+ - Extract the cookie for the current display
+ - Create a new database file
+ - Add an entry for the spoofed display using the extracted cookie to the new
+   database
+ - Rewrite the hostname of the entry that was just created to be "FamilyWild"
```

Comment:
What's the significance of the FamilyWild name?

Reply:
It's the term the specs for `xauth` use.

https://gitlab.freedesktop.org/xorg/lib/libxau

> The Family is specified in two bytes to allow out-of-band values
(i.e. values not in the Protocol) to be used.  In particular,
two new values "FamilyLocal" and "FamilyWild" are defined.  FamilyLocal
refers to any connections using a non-network method of connection from the
local machine (Unix domain sockets, shared memory, loopback serial line).
In this case the host address is specified by the data returned from
gethostname() and better be unique in a collection of machines
which share NFS directories.  FamilyWild is currently used only
by xdm to communicate authorization data to the server.  It matches
any family/host address pair.